Support GRIB2 rotated latitude/longitude grid as CF Conventions rotat…

bencaradocdavies opened the pull request:
…ed pole grid

[NOAA Rapid Refresh (RAPv3)](http://rapidrefresh.noaa.gov/) native grid GRIB2 files use a [rotated latitude/longtude grid with GDS template 32769](http://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_temp3-32769.shtml).

This pull request:
* adds support for GDS template 32769 grids and,
* fixes support for GDS template 1 grids, and
* uses ```RotatedPole``` projection (not ```RotatedLatLon```), and
* encodes the projection as a [CF Conventions rotated pole grid](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch05s06.html).

Strictly speaking, the implementation still lacks the ```lat(rlat,rlon)``` and ```lon(rlat,rlon)``` variables required by the CF Conventions, but this implementation is much closer than the previous version which bore no resemblance to the CF Conventions. Note also that these additional variables confuses some consumers which give priority to ```lat``` and ```lon``` and fail if they are multidimensional.

Note that #530 is also required to consume RAPv3 native grid GRIB2 files.

I am happy to port this to 5.0.0 if you like.

bencaradocdavies commented:
```Double.hashCode(double)```: new in Java 8. Will fix.

braeckel commented:
We would find this very useful.  +1 for integrating this improvement

cwardgar commented:
We had some failures on Jenkins for this one: https://jenkins-aws.unidata.ucar.edu/job/cwardgar/18/testReport/

I'm seeing a lot of `Unexpected nonpositive upper right rotated latitude`.

bencaradocdavies commented:
@cwardgar is there anywhere I can get these two test data files?:
```
/share/testdata/cdmUnitTest/formats/grib2/COSMO_EU.grib2
/share/testdata/cdmUnitTest/formats/grib2/cosmo_de_eps_m001_2009051100.grib2
```
I am flying blind on GDS template 1 support as I have no test data. I know that Unidata have an extensive private test suite and I am pleased to see it in action.

bencaradocdavies commented:
@braeckel what GDS template is of most interest to you? GDS template 1 or the RAPv3 native grid template 32769?

bencaradocdavies commented:
I see that I used the wrong value in all the latitude exception messages (```getX()``` should be ```getY()```) , but the correct value is used in the conditions. I will have to debug against the test data files, when I have them.

cwardgar commented:
@bencaradocdavies Here you go: ftp://ftp.unidata.ucar.edu/pub/cwardgar/gh-531/

bencaradocdavies commented:
Thanks very much @cwardgar, I am downloading now.

Googling indicated that there were some problems with the projection for COSMO_EU.grib2:
https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg09462.html

I will compare master and my pull request in toolsUI to check that I am not breaking something that works for GDS template 1. In my view, the implementation was broken, but it might be better to leave it as is if it is being used by clients, and confine my changes to GDS template 32789.

bencaradocdavies commented:
@cwardgar here is the broken projection for COSMO_EU reported by John Caron in August 2010:
![tor.png](https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/png9bJFdblWeN.png)
From: https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg09462.html

Here is what I see on master (before this PR):
![cosmo_eu_toolsui_master](https://cloud.githubusercontent.com/assets/1900866/14546805/fc47f4d6-02fb-11e6-81fc-89155b17e44b.png)

This PR is still broken, but now I am happier about changing the behaviour for GDS template 1.

bencaradocdavies commented:
And the smaller test file gives the same result, centred on latitude/longitude zero. This means that COSMO is encoding la1/lo1/la2/lo2 as the *rotated* grid corners, not unrotated. I will investigate.

bencaradocdavies commented:
@cwardgar I have updated the request. The main change is to not reproject la1/lo1/la2/lo2 for GDS template 1 as the test data gives these in the *rotated* coordinate system. This differs from GDS template 32769 in which la1/lo1/la2/lo2 are in the unrotated coordinate system (lower left and centre respectively). I also added lon/lat normalisation for consistency with the projection information and to make them play nicely with ToolsUI.

With this pull request, I see in ToolsUI FeatureTypes / Grid Viewer the two GDS template 1 files now centred on Germany and the GDS template 32769 file I have centred on the contiguous United States:

![cosmo_eu](https://cloud.githubusercontent.com/assets/1900866/14548918/7809ad4a-030f-11e6-9c7c-b524bbb67539.png)

![cosmo_de_eps_m001_2009051100](https://cloud.githubusercontent.com/assets/1900866/14548925/821d5dd6-030f-11e6-83e6-60b9a20da188.png)

![rap-native-grid](https://cloud.githubusercontent.com/assets/1900866/14548927/894d1c2c-030f-11e6-9b16-8c19f2d8f26b.png)

cwardgar commented:
[Here](https://jenkins-aws.unidata.ucar.edu/job/cwardgar/19/testReport/) is the Jenkins result on the latest commit. Looks good! Those 4 failures you see [are unreleated](https://github.com/Unidata/thredds/issues/193). @lesserwhirls, any comment before we merge this in?

lesserwhirls commented:
I think we're good. @bencaradocdavies - are you still willing to give porting to v5.0 a shot?

bencaradocdavies commented:
Thanks very much @cwardgar and @lesserwhirls for your prompt attention to this PR. I will see if I can port this (and #530) to 5.0.0.
